proposals/csharp-10.0/file-scoped-namespaces.md

```diff
@@ -0,0 +1,129 @@
+
+### Motivation
+
+Analysis of the C# ecosystem shows that around 99.7% (or more) files are all of either one of these forms:
+
+```c#
+namespace X.Y.Z
+{
+    // usings
+
+    // types
+}
+```
+
+or
+
+```c#
+// usings
+
+namespace X.Y.Z
+{
+    // types
+}
+```
+
+However, both these forms force the user to indent the majority of their code and add a fair amount of ceremony for what is effectively a very basic concept.  This affects clarity, uses horizontal and vertical space, and is often unsatisfying for users both used to C# and coming from other languages (which commonly have less ceremony here).
+
+The primary goal of the feature therefor is to meet the needs of the majority of the ecosystem with less unnecessary boilerplate.
+
+### Detailed design
+
+This proposal takes the form of a diff to the existing https://github.com/dotnet/csharplang/blob/main/spec/namespaces.md#compilation-units section of the specification.
+
+#### Diff
+
+~~A *compilation_unit* defines the overall structure of a source file. A compilation unit consists of zero or more *using_directive*s followed by zero or more *global_attributes* followed by zero or more *namespace_member_declaration*s.~~
+
+A *compilation_unit* defines the overall structure of a source file. A compilation unit consists of zero or more *using_directive*s followed by zero or more *global_attributes* followed by a *compilation_unit_body*. A *compilation_unit_body* can either be a *file_scoped_namespace_declaration* or zero or more *statement*s and *namespace_member_declaration*s.
+
+```antlr
+compilation_unit
+~~    : extern_alias_directive* using_directive* global_attributes? namespace_member_declaration*~~
+    : extern_alias_directive* using_directive* global_attributes? compilation_unit_body
+    ;
+
+compilation_unit_body
+    : statement* namespace_member_declaration*
+    | file_scoped_namespace_declaration
+    ;
+```
+
+... unchanged ...
+
+A *file_scoped_namespace_declaration* will contribute members corresponding to the *namespace_declaration* it is semantically equivalent to.  See ([Namespace Declarations](#namespace-declarations)) for more details.
+
+## Namespace declarations
+
+A *namespace_declaration* consists of the keyword `namespace`, followed by a namespace name and body, optionally followed by a semicolon.
+A *file_scoped_namespace_declaration* consists of the keyword `namespace`, followed by a namespace name, a semicolon and an optional list of *extern_alias_directive*s, *using_directive*s and *type_declaration*s.
+
+```antlr
+namespace_declaration
+    : 'namespace' qualified_identifier namespace_body ';'?
+    ;
+    
+file_scoped_namespace_declaration
+    : 'namespace' qualified_identifier ';' extern_alias_directive* using_directive* type_declaration*
```

Comment:
>extern_alias_directive* using_directive* type_declaration* [](start = 43, length = 58)

I'm surprised the `file_scoped_namespace_declaration` contains the directives and type declarations that follow. Is that necessary, and does that mean the corresponding syntax nodes would have the same nesting?

Reply:
the purpose of the grammar here is to just define strictly what people can and cannot write.  In this case, i found this to be hte simplest and clearest way to write the grammar to achieve the design we wanted.

> and does that mean the corresponding syntax nodes would have the same nesting?

As it turns out, yes.  My approach here (which is similar towhat we did in typescript) is that there is now a BaseNamespaceDeclaration, which has hte name, and members. And then there are two subclasses representing the two different syntactic forms.  This approach makes the implementation (for both the IDE and compiler) extremely trivial as the same 'tree shape' (esp. wrt to parent/children nodes) is present for both trees.

Reply:
> My approach here (which is similar towhat we did in typescript) is that there is now a BaseNamespaceDeclaration, which has hte name, and members. And then there are two subclasses representing the two different syntactic forms.

But it doesn't feel like there is a syntactic relationship between the `namespace identifier;` and the subsequent directives and types.

The only analogy I can think of to `namespace identifier;` in terms of how it applies to subsequent code is a `using var identifier = expr;`, and the statements following the using statement are not children of using statement syntax node.

---
In reply to: [601267615](https://github.com/dotnet/csharplang/pull/4581#discussion_r601267615) [](ancestors = 601267615)

Reply:
> But it doesn't feel like there is a syntactic relationship between the namespace identifier; and the subsequent directives and types.

It's going to come down to perspective. Either approaches are valid, however one is vastly more costly. For example, every piece of code that walks upwards to determine what namespace it is in must now be revised to not only do the standard walk, but also then got the compilation unit *and* then check if there is one of these directives. 

As mentioned earlier, we took this approach in ts with great success. We could take this other approach but it will mean a higher cost for both codebases.